typescript/schedules.md

```diff
@@ -0,0 +1,1255 @@
+    identity: `${process.pid}@${os.hostname()}`,
+    interceptors: {},
+    namespace: 'default',
+  };
+}
+
+// TODO
+// function ensureArgs<W extends Workflow, T extends ScheduleOptions<W>>(opts: T): Omit<T, 'args'> & { args: unknown[] } {
+//   const { args, ...rest } = opts;
+//   return { args: args ?? [], ...rest };
+// }
+
+interface SchdeduleHandleOptions extends GetSchdeduleHandleOptions {
+  workflowId: string;
+  runId?: string;
+  interceptors: ScheduleClientCallsInterceptor[];
+  /**
+   * A runId to use for getting the workflow's result.
+   *
+   * - When creating a handle using `getHandle`, uses the provided runId or firstExecutionRunId
+   * - When creating a handle using `start`, uses the returned runId (first in the chain)
+   * - When creating a handle using `signalWithStart`, uses the the returned runId
+   */
+  runIdForResult?: string;
+}
+
+/**
+ * Policy for overlapping Actions.
+ */
+export enum ScheduleOverlapPolicy {
+  /**
+   * Don't start a new Action.
+   */
+  Skip = 1,
+
+  /**
+   * Start another Action as soon as the current Action completes, but only buffer one Action in this way. If another
+   * Action is supposed to start, but one Action is running and one is already buffered, then only the buffered one will
+   * be started after the running Action finishes.
+   */
+  BufferOne,
+
+  /**
+   * Allows an unlimited number of Actions to buffer. They are started sequentially.
+   */
+  BufferAll,
+
+  /**
+   * Cancels the running Action, and then starts the new Action once the cancelled one completes.
+   */
+  CancelOther,
+
+  /**
+   * Terminate the running Action and start the new Action immediately.
+   */
+  TerminateOther,
+
+  /**
+   * Allow any number of Actions to start immediately.
+   *
+   * This is the only policy under which multiple Actions can run concurrently.
+   */
+  AllowAll,
+}
+
+checkExtends<temporal.api.enums.v1.schedule.ScheduleOverlapPolicy, ScheduleOverlapPolicy>();
+
+export interface Backfill {
+  start: Date;
+  end: Date;
+  /**
+   * Override overlap policy for this request.
+   */
+  overlap?: ScheduleOverlapPolicy;
+}
+
+/**
+ * The range of values depends on which, if any, fields are set:
+ *
+ * ```
+ * {} -> all values
+ * {start} -> start
+ * {start, end} -> every value from start to end (implies step = 1)
+ * {start, step} -> from start, by step (implies end is max for this field)
+ * {start, end, step} -> from start to end, by step
+ * {step} -> all values, by step (implies start and end are full range for this field)
+ * {end} and {end, step} are not allowed
+ * ```
+ * TODO is there a way to express not allowed ^ in type?
+ *
+ * For example:
+ *
+ * ```
+ * {start: 2, end: 10, step: 3} -> 2, 5, 8
+ * ```
+ */
+export interface Range<Unit> {
+  start?: Unit;
+  end?: Unit;
+
+  /**
+   * The step to take between each value.
+   *
+   * @default 1
+   */
+  step?: PositiveInteger;
+}
+
+export type AnyValue = '*';
+
+export type ZeroTo59 =
```

Comment:
This and all of the similar types like this are silly IMO. Just take number and validate at runtime. If you only accept these literals here, does this make it annoying for people that, say, accept this value from config file or user input or some other numerical place.

Reply:
Yes, they'd have to validate. I guess it's easier for them to handle our validation errors than to implement themselves

Reply:
What I mean is we should validate at runtime. You don't need to try to validate in the type system at this low of a level IMO.

Reply:
Yeah, I meant them handle the validation errors we throw at runtime